bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java

```diff
@@ -860,41 +862,45 @@ private void teardown()
       }
       torndown = true;
 
-      /*
-       * if the publisher unregistered, all the downstream guys are going to be unregistered anyways
-       * in our world. So it makes sense to kick them out proactively. Otherwise these clients since
-       * are not being written to, just stick around till the next publisher shows up and eat into
-       * the data it's publishing for the new subscribers.
-       */
-
-      /**
-       * since the publisher server died, the queue which it was using would stop pumping the data unless
-       * a new publisher comes up with the same name. We leave it to the stream to decide when to bring up a new node
-       * with the same identifier as the one which just died.
-       */
-      if (publisherChannels.containsValue(this)) {
-        final Iterator<Entry<String, AbstractLengthPrependerClient>> i = publisherChannels.entrySet().iterator();
-        while (i.hasNext()) {
-          if (i.next().getValue() == this) {
-            i.remove();
-            break;
+      serverHelperExecutor.submit(() ->
+      {
+        /*
+         * if the publisher unregistered, all the downstream guys are going to be unregistered anyways
+         * in our world. So it makes sense to kick them out proactively. Otherwise these clients since
+         * are not being written to, just stick around till the next publisher shows up and eat into
+         * the data it's publishing for the new subscribers.
+         */
+
+        /**
+         * since the publisher server died, the queue which it was using would stop pumping the data unless
+         * a new publisher comes up with the same name. We leave it to the stream to decide when to bring up a new node
+         * with the same identifier as the one which just died.
+         */
+        if (publisherChannels.containsValue(this)) {
```

Comment:
`publisherChannels.values.remove(Publisher.this)`?

Reply:
values.remove seems to be instantiating a new view internally, creating iterators and iterating through the map similar to iteration being done here (the if containsValue can be removed as that is one more traversal). There is extra memory overhead with no gain in performance and possibly only improves readability. I prefer to just remove the if and leave the rest as is.

Reply:
The extra memory is one reference and it saves us from maintaining code that is already part of JDK :)

Reply:
You mean one new object right ;)? It's just that creating a values view just for removing something feels wrong.

Reply:
Yes, it is one new instance of an object that holds a single reference to the original `publisherChannels` map.

Reply:
Actually, can you do `publisherChannels.remove(Publisher.this.datalist.getIdentifier(), Publisher.this);`?

Reply:
Is there a chance the datalist is not yet assigned to the publisher by the time teardown gets called?

Reply:
I worry this will introduce some uncertainty without enough real-world testing. If it is bothering you much, I will change it to your earlier suggestion of removing from values.

Reply:
`datalist` can't be `null`. It is `final`.